core/src/avm2/globals/json.rs

```diff
@@ -0,0 +1,443 @@
+    /// to the replacer function with the key in a (key, value) pair, and the value is mapped to the return value
+    /// of the replacer function. If the user did not supply a replacer function, this step is skipped.
+    ///
+    /// The `key` is lazily evaluated because it may be expensive in some areas to generate the key, but the key is
+    /// only used if either the `toJSON` step or replacer function step happens, so we only need to evaluate the key there.
+    fn map_value(
+        &self,
+        activation: &mut Activation<'_, 'gc, '_>,
+        key: impl Fn() -> AvmString<'gc>,
+        value: Value<'gc>,
+    ) -> Result<Value<'gc>, Error> {
+        let (eval_key, value) = if value.is_primitive() {
+            (None, value)
+        } else {
+            let obj = value.coerce_to_object(activation)?;
+            let to_json = obj
+                .get_property(
+                    obj,
+                    &QName::new(Namespace::public(), "toJSON").into(),
+                    activation,
+                )?
+                .coerce_to_object(activation)
+                .ok()
+                .and_then(|obj| obj.as_function_object());
+            if let Some(to_json) = to_json {
+                let key = key();
+                (Some(key), to_json.call(None, &[key.into()], activation)?)
+            } else {
+                (None, value)
+            }
+        };
+        if let Some(Replacer::Function(replacer)) = self.replacer {
+            replacer.call(
+                None,
+                &[eval_key.unwrap_or_else(key).into(), value],
+                activation,
+            )
+        } else {
+            Ok(value)
+        }
+    }
+
+    fn serialize_object(
+        &mut self,
+        activation: &mut Activation<'_, 'gc, '_>,
+        obj: Object<'gc>,
+    ) -> Result<JsonValue, Error> {
+        let mut js_obj = JsonObject::new();
+        // If the user supplied a PropList, we use that to find properties on the object.
+        if let Some(Replacer::PropList(props)) = self.replacer {
+            let mut iter = ArrayIter::new(activation, props.into())?;
+            while let Some(r) = iter.next(activation) {
+                let item = r?.1;
+                let key = item.coerce_to_string(activation)?;
+                let value = obj.get_property(
+                    obj,
+                    &QName::new(Namespace::public(), key).into(),
+                    activation,
+                )?;
+                let mapped = self.map_value(activation, || key, value)?;
+                if !matches!(mapped, Value::Undefined) {
+                    js_obj.insert(
+                        &key.to_utf8_lossy(),
+                        self.serialize_value(activation, mapped)?,
+                    );
+                }
+            }
+        } else {
+            for i in 1.. {
+                // TODO: We should get more than just enumerable properties
```

Comment:
Some extra "fun" cases:
stringifying a function appears to produce `null`
stringifying a class object 
```
class C {
    public static var X = 1;
}
```
appears to produce `{"X":1,"prototype":{}}`

Also, in general, it does indeed iterate more than just enumerable properties.
```
dynamic class C {
    public var X = 1;
    public var Y = 2;
}
var a = new C();
a.Z = 3;

for (var x in a) {
    trace(x);
}

trace(JSON.stringify(a));
```
produces
```
Z
{"X":1,"Y":2,"Z":3}
```

BTW this might be more consistent to write after my refactor PR, as I introduce a clean split between non-dynamic (trait) and dynamic instance properties. But I think it'd also me okay if this got merged first and improved later, as long as we have tests ready.